### PR TITLE
K8SPXC-730: Rework statuses for the Custom Resource

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPXC-730](https://jira.percona.com/browse/K8SPXC-730) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

This commit introduces two new statuses:
1. Stopping (cluster is being paused or deleted)
2. Paused (cluster is paused *completely*)

### CODE REVIEW

File: pkg/apis/pxc/v1/pxc_types.go [hunk 1]
```diff
@@ -893,10 +895,14 @@ func (cr *PerconaXtraDBCluster) ProxySQLEnabled() bool {
...
 }
 
-func (s *PerconaXtraDBClusterStatus) ClusterStatus(inProgress bool) AppState {
+func (s *PerconaXtraDBClusterStatus) ClusterStatus(inProgress, deleted bool) AppState {
 	switch {
 	case s.PXC.Status == AppStateError || s.ProxySQL.Status == AppStateError || s.HAProxy.Status == AppStateError:
 		return AppStateError
+	case deleted || s.PXC.Status == AppStateStopping || s.ProxySQL.Status == AppStateStopping || s.HAProxy.Status == AppStateStopping:
+		return AppStateStopping
+	case s.PXC.Status == AppStatePaused && (s.ProxySQL.Status == AppStatePaused || s.HAProxy.Status == AppStatePaused):
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:904]: what if both ProxySQL and HAProxy are disabled?

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| upgrade-haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| proxysql-sidecar-res-limits | passed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | passed |
| users | passed |
| demand-backup | passed |
| security-context | passed |
| pitr | passed |
| storage | passed |
| scheduled-backup | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| self-healing-advanced | passed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/894/commits/74f1412aa5221f836d0f4447c0085b81e59ca9a5
image: `perconalab/percona-xtradb-cluster-operator:PR-894-74f1412a`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| scheduled-backup | passed |
| init-deploy | passed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring-2-0 | passed |
| smart-update | failed |
| affinity | passed |
| one-pod | failed |
| upgrade-consistency | passed |
| auto-tuning | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/894/commits/2de8362eea032690c3369e71965c5dbc3fc74890
image: `perconalab/percona-xtradb-cluster-operator:PR-894-2de8362e`

File: pkg/apis/pxc/v1/pxc_types.go [hunk 2]
```diff
@@ -895,16 +897,16 @@ func (cr *PerconaXtraDBCluster) ProxySQLEnabled() bool {
...
-	case s.PXC.Status == AppStateError || s.ProxySQL.Status == AppStateError || s.HAProxy.Status == AppStateError:
-		return AppStateError
-	case inProgress || s.PXC.Status == AppStateInit || s.ProxySQL.Status == AppStateInit || s.HAProxy.Status == AppStateInit:
-		return AppStateInit
-	case s.PXC.Status == AppStateReady:
+	case deleted || s.PXC.Status == AppStateStopping || s.ProxySQL.Status == AppStateStopping || s.HAProxy.Status == AppStateStopping:
+		return AppStateStopping
+	case s.PXC.Status == AppStatePaused && ((s.ProxySQL.Status == AppStatePaused || s.HAProxy.Status == AppStatePaused) || true):
+		return AppStatePaused
+	case s.PXC.Status == AppStateReady && ((s.ProxySQL.Status == AppStateReady || s.HAProxy.Status == AppStateReady) || true):
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:906]: `((s.ProxySQL.Status == AppStatePaused || s.HAProxy.Status == AppStatePaused) || true)` condition is always true, it impossible to get false from it

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/pxc/v1/pxc_types.go (hunk 2 above)
Comment (egegunes) [reply pkg/apis/pxc/v1/pxc_types.go]: fixed in https://github.com/percona/percona-xtradb-cluster-operator/pull/894/commits/650a935ad7d1abf5850c129ac42584562e06c069

File: pkg/apis/pxc/v1/pxc_types.go (hunk 1 above)
Comment (egegunes) [reply pkg/apis/pxc/v1/pxc_types.go]: fixed in https://github.com/percona/percona-xtradb-cluster-operator/pull/894/commits/650a935ad7d1abf5850c129ac42584562e06c069

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | passed |
| big-data | passed |
| storage | passed |
| haproxy | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| upgrade-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| smart-update | failed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| auto-tuning | passed |
| upgrade-consistency | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| pitr | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/894/commits/650a935ad7d1abf5850c129ac42584562e06c069
image: `perconalab/percona-xtradb-cluster-operator:PR-894-650a935a`

**Review by delgod — APPROVED** [review]

File: pkg/controller/pxc/status.go [hunk 3]
```diff
@@ -102,16 +101,9 @@ func (r *ReconcilePerconaXtraDBCluster) updateStatus(cr *api.PerconaXtraDBCluste
 
 		cr.Status.Size += status.Size
 		cr.Status.Ready += status.Ready
-
-		if !inProgress {
-			inProgress, err = r.upgradeInProgress(cr, a.app.Name())
-			if err != nil {
-				return errors.Wrapf(err, "check %s upgrade progress", a.app.Name())
-			}
```
[the remark this answers is not in the document]
Comment (egegunes) [reply pkg/controller/pxc/status.go:110]: we don't, default status is `initializing`

[the remark this answers is not in the document]
Comment (delgod) [reply pkg/controller/pxc/status.go:110]: so, if upgradeInProgress function not needed anymore, please remove it.
also integrate logic from this function somewhere here. we should not show "Ready" status if `UpdatedReplicas < Replicas`

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| haproxy | passed |
| scaling | passed |
| recreate | passed |
| big-data | passed |
| scheduled-backup | passed |
| storage | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| affinity | passed |
| smart-update | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| tls-issue-self | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/894/commits/c4b45c8e1529fc5c46574e3686ff117295cbf263
image: `perconalab/percona-xtradb-cluster-operator:PR-894-c4b45c8e`

**Review by delgod — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Pull request merged**